components/core/src/clp/ffi/SchemaTree.hpp

```diff
@@ -0,0 +1,151 @@
+#ifndef CLP_FFI_SCHEMATREE_HPP
+#define CLP_FFI_SCHEMATREE_HPP
+
+#include <cstddef>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <utility>
+#include <vector>
+
+#include "../ErrorCode.hpp"
+#include "../TraceableException.hpp"
+#include "SchemaTreeNode.hpp"
+
+namespace clp::ffi {
+/**
+ * This class implements a schema tree, providing all necessary methods for upper-layer APIs through
+ * FFI. Nodes are stored using a flattened vector.
+ */
```

Comment:
Since clp_s::SchemaTree is undocumented, let's take this opportunity to document this class.

```suggestion
/**
 * A tree representing the schema of a set of JSON (or any format with a dynamic structure) records.
 * Each node in a SchemaTree has a key name and a value type, corresponding to the key name and
 * value type of a key-value pair (kv-pair) in a record. A nested kv-pair is represented as a child
 * node of the outer kv-pair. As a result, all leaf nodes represent primitive values while internal
 * nodes represent non-primitive values.
 *
 * NOTE:
 * - The tree is the representation of multiple record's schemas merged together, so technically, it
 *   should be called a MergedSchemaTree; we use SchemaTree for simplicity.
 * - The root node is special with key name "" and type `SchemaTreeNode::Type::Obj`. This means that
 *   this SchemaTree implementation cannot represent records that are not objects.
 * - This SchemaTree implementation does not represent arrays. Instead, arrays in a record can be
 *   serialized as strings and represented using the type `SchemaTreeNode::Type::UnstructuredArray`.
 *
 * For example, consider the two records below.
 *
 * # Record 1
 * {
 *   "a": 0,
 *   "b": {
 *     "c": true
 *   }
 * }
 *
 * # Record 2
 * {
 *   "a": 0.0,
 *   "d": {
 *     "e": "E"
 *   }
 * }
 *
 * The schema trees for each record are below.
 *
 * # Record 1's schema tree
 * <: Obj> --> <a: Int>
 *         --> <b: Obj>   --> <c: Bool>
 *
 * # Record 2's schema tree
 * <: Obj> --> <a: Float>
 *         --> <d: Obj>   --> <e: Str>
 *
 * In the diagrams above, a node is represented by <KeyName: Type> and a parent-child relationship
 * is represented using "-->".
 *
 * When the records' schema trees are merged together, they create the following SchemaTree:
 *
 * <: Obj> --> <a: Int>
 *         --> <a: Float>
 *         --> <b: Obj>   --> <c: Bool>
 *         --> <d: Obj>   --> <e: Str>
 *
 * Notice that nodes with the same key name, type, and parents are merged together. Nodes that
 * differ in just their key name (e.g., "a") remain unique.
 */
```